share/wake/lib/core/list.wake

```diff
@@ -845,19 +844,99 @@ export def groupBy (cmp: a => a => Boolean) (list: List a): List (List a) =
+# Given a list of pairs, find all right elements such that
+# their left element matches a selection function. This
+# can be used to lookup values by keys if the list
+# represents a dictionary for instance.
+#
+# Parameters:
+#  - `selectFn`: The function that selects which keys to filter by
+#  - `list`: The list of key-value pairs to pull selected values out of
+#
+# Examples:
+# ```
+#   def l = ("Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil)
+#   lookup (_==~"Alice") l = 85, Nil
+#   lookup (_==~"Bob") l = 4, 15, Nil
+#   lookup (_==~"Mal") l = Nil
+#   lookup (\_ True) l = 85, 4, 20, 15, Nil
+#   lookup (\_ False) l = Nil
+# ```
+export def lookup (selectFn: a => Boolean): (list: List (Pair a b)) => List b =
   def select (Pair key value) = if selectFn key then Some value else None
   mapPartial select
 
-export def editLookup (selectFn: a => Boolean) (editFn: b => b): List (Pair a b) => List (Pair a b) =
+# Given a list of pairs, edit all right elements such that
+# their left element matches a selection function. This
+# can be used to eidts values in a dictionary for instance.
+#
+# Parameters:
+#  - `selectFn`: The function that selects which keys to edit the values of
+#  - `editFn`: The function used to update values
+#  - `list`: The list of key-value pairs to pull selected values out of
+#
+# Examples:
+# ```
+#   def l = ("Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil)
+#   editLookup (_==~"Alice") (_-5) l = "Alice" :-> 80, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil
+#   editLookup (_==~"Bob") (_-5) l = "Alice" :-> 85, "Bob" :-> -1, "Carol" :-> 20, "Bob" :-> 10, Nil
+#   editLookup (_==~"Mal") (_+1000) l = "Alice" :-> 85, "Bob" :-> 4, "Carol" :-> 20, "Bob" :-> 15, Nil
+#   editLookup (\_ True) (_+1) l = "Alice" :-> 86, "Bob" :-> 5, "Carol" :-> 21, "Bob" :-> 16, Nil
+#   editLookup (\_ False) (+1) l = "Alice" :-> 86, "Bob" :-> 5, "Carol" :-> 21, "Bob" :-> 16, Nil
+# ```
+export def editLookup (selectFn: a => Boolean) (editFn: b => b): (list: List (Pair a b)) => List (Pair a b) =
   editFilter (selectFn _.getPairFirst) $ editPairSecond $ editFn
 
+# Checks to see if `list` contains exactly one element and
+# either returns that element or `msg` as an error otherwise.
+# This can be used to close off chains of list operations where
+# you expect exactly one element at the end.
+#
+# Parameters:
+#  - `msg`: The message of the error if the list does not have exactly one element
+#  - `list`: The input list thats expected to have exactly one element.
+#
+# Examples:
+# ```
+#   expectOne "foo" (1, Nil)    = Pass 1
+#   expectOne "foo" Nil         = Fail (Error "foo" ...)
+#   expectOne "foo" (1, 2, Nil) = Fail (Error "foo" ...)
+#   seq 10 | filter isPrime | filter isEven | expectOne "you broke math!" = Pass 2
+#   seq 10 | filter isPrime | filter isOdd | expectOne "too many odd primes" = Fail (Error "too many odd primes" ...)
+# ```
```

Comment:
Seeing the examples here makes me realize how much I dislike this function. You want two different error messages in the underflow and overflow case.

When I've done this in the past, I've used the `require/match` pattern:

```
require Pass foo = match list
  x, Nil = Pass x
  x, _   = failWithError "too many"
  Nil    = failWithError "too few"
```

Reply:
It's a common pattern and it might be nice to simplify a bit, but now that you point it out, I'm not even sure we'd want to have the context-append; most of the time, the reason for a `Nil` list is going to be rather different from that for an overlong list, and even if it isn't it takes a bit extra mental bandwidth to remember that the context will need to fit both cases and then come up with a satisfactory wording.  The only real way I can see `expectOne` working is if it took two separate message strings, and at that point we may as well be back at the `match`.

If people want the `expect*` functions, they're easy enough to add to a project-specific common package.  At this point, I don't think we need any of them in core.

Reply:
Yeah I hit that same gripe when writing these examples so sounds good. I'll remove them.

Reply:
Yeah I hit this exact issue while writing these examples. And I agree it is small enough for users to implement it themselves and I don't think we want to add two parameters to expectOne to resolve this. I'll remove the expect functions but I do think we can do better about error handling here somehow though. I'll remove these for now and think about it.

My primary issue with using lists is that in *most* cases you'll see a 0/1 case rather than a 0/1/more than 1 case. Mentioning List constructors in patterns really breaks with the abstraction that we've built up here IMO and brings us further from how this might be handled in more familiar languages (though "expectOne" isn't really how this would be handled outside of a language like Rust which has a very steep learning curve)

Reply:
I'd argue it only breaks the abstraction if we didn't want to reveal that the query language maintains a running set of candidate matches. I'd argue that this is explicitly what `jq` does, and we do too. We also make this explicit promise because we want to support `filter` / `catWith` / etc on the results.

Reply:
Actually, we probably don't even want to promise JUST a set. There is a defined (and useful) order to the candidate matches. So, at most we could have been generalizing of some sort of `Seq` type-class (that includes Vector/List/Option). Unfortunately, since we don't have that, `List` is what we've pretty clearly selected and exposed.

Reply:
Yeah for things like JArray order is critical.

Really my complaint here is that I want both Option and List and that's not really something we can easily get in Wake right now (nor is there any design on the horizon that would allow it to work well). Something something "typeclasses would fix this" lol